### ISSUE TITLE
Enabling --ipip on a pool doesn't have desired effect if "node" is already started

### ISSUE THREAD

**Issue by tomdee** [projectcalico/calico-containers#716]
I was trying out calico on GCE and ran `calicoctl node` on all my hosts. I then realised that I'd need --ipip so I ran `calicoctl pool add x.x.x.x/x --ipip` to update my pool.

I created some containers but found they couldn't communicate with each other. The issue was revealed by looking at the routing table

```
default via 10.240.0.1 dev ens4v1  proto dhcp  src 10.240.0.5  metric 1024 
default via 10.240.0.1 dev ens4v1  proto dhcp  metric 1024 
10.240.0.1 dev ens4v1  proto dhcp  scope link  src 10.240.0.5  metric 1024 
10.240.0.1 dev ens4v1  proto dhcp  metric 1024 
172.17.0.0/16 dev docker0  proto kernel  scope link  src 172.17.0.1 
192.168.0.0/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.0.64/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.0.128/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.0.192/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.1.0/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.1.64/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.1.128/26 via 10.240.0.1 dev ens4v1  proto bird 
blackhole 192.168.1.192/26  proto bird 
192.168.1.213 dev cali37477112b4a  scope link 
192.168.2.0/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.2.64/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.2.128/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.2.192/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.3.0/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.3.64/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.3.128/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.3.192/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.4.0/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.4.64/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.4.128/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.4.192/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.5.0/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.5.64/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.5.128/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.5.192/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.6.0/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.6.64/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.6.128/26 via 10.240.0.1 dev ens4v1  proto bird 
192.168.6.192/26 via 10.240.0.1 dev ens4v1  proto bird 

```

Some part of confd or bird hasn't spotted that ipip was enabled and so all the routes were pointing at the wrong place (the gateway instead of directly at the host)

**Comment by fasaxc** [comment]
I hit this too. I tried stopping and starting my calico/node but that didn't resolve it.  I had to reboot to fix it.

**Comment by fasaxc** [comment]
I was testing the Ubuntu k8s set up guides that I've been working on.  As part of that, I switched the guides to use 0.13.0, in case that's relevant.

CC @caseydavenport who has the details of what I was testing, in case that's needed while I'm OOF.

**Comment by matthewdupre** [comment]
Candidate fix: flush all the BIRD programmed routes when changing IPIP.

**Comment by matthewdupre** [comment]
A little bit of additional information.

- This looks like some kind of timing issue between setup of the IPIP tunnel device tunl0 and BIRD picking up the config for it.
- The problem only happens once: if you change from no-ipip to ipip then the routes won't be fixed, but if you later restart the BIRDs you can't reproduce on the same system by toggling back to no-ipip (this is a pain!).  From this point forwards the changes work nicely.
- This is probably related to the fact that once Felix has loaded the kernel module to create the tunnel, the tunnel is kept around (even surviving reboots) whether IPIP is enabled on the pool or not.

**Comment by matthewdupre** [comment]
Possible fix approaches:
- (partial) delay BIRD restart (perhaps full restart not HUP),
- see if there's a problem with the krt_tunnel code in BIRD.

I'll take a quick look at the BIRD patches to see if there's a smoking gun.

**Comment by matthewdupre** [comment]
Also node that turning IPIP on, then off then on again also works (essentially giving some time for the tunl0 to be established).

**Comment by aleks-v-k** [comment]
I have a similar problem with kubernetes & calico (as described in duplicates of this issue: #772 & #947 ).

After rebooting a node most of the time I have broken routes on that node:
```
$ diff broken/diagnostics/ip_-4_route worked/diagnostics/ip_-4_route
2,3c2,3
< 10.1.0.0/26 via 192.168.57.51 dev enp0s3  proto bird 
< 10.1.0.64/26 via 192.168.57.52 dev enp0s3  proto bird 
---
> 10.1.0.0/26 via 192.168.57.51 dev tunl0  proto bird onlink 
> 10.1.0.64/26 via 192.168.57.52 dev tunl0  proto bird onlink
```
If I delete a wrong route (with enp0s3 interface), then there will be added new
one with correct route (tunl0 interface) and access between pods on different
nodes will work.

I have two questions:

1. When you are planning to fix this issue?
2. What a workaround would you suggest before the fix will be available?

**Comment by caseydavenport** [comment]
@matthewdupre what's the status on this guy?  @djosborne is this related to the recent `ipip` fix you made?

**Comment by djosborne** [comment]
The `ipip` change I made no longer requires `calico/node`  be restarted when you add your first pool with `--ipip`. It uses confd to watch pools, and when one with `ipip` is added, it [does what node does at startup](https://github.com/projectcalico/calico-containers/blob/2d0b6d5d15a5556742f7af3cc076b00dcfcf6f11/calico_node/filesystem/startup.py#L70) - it allocates one IP from that pool and stores it as the host's `IpInIpTunnelAddr` in etcd. 

My fix is in the same ballpark, but at first glance it doesn't seem like it would fix or worsen this issue.

**Comment by matthewdupre** [comment]
Thanks Alex: adding a little detail / some possible narrowing of where the fix is needed.
- the route in question was `<Calico /26> via <Remote host> dev ethX proto bird onlink`
- it needed to be `<Calico /26> via <Remote host> dev tunl0 proto bird onlink`
- Since BIRD was observed reprogramming it correctly, it seems very likely that any fix needed should be in BIRD, or around it.

I guess we could either modify BIRD to properly reprogram this route when it gets the krt_tunnel attribute; or maybe just restart BIRD when enabling IPIP (easier, but I'd be worried about possible timing issues).

**Comment by caseydavenport** [comment]
> or maybe just restart BIRD when enabling IPIP (easier, but I'd be worried about possible timing issues

I'd like to go down this path if we could since it should be simpler, but there is definitely concern for restarting all BIRD instances across a cluster at once.  We enable graceful restart, so we should be able to make this work without affecting the data plane.  We likely just need to add some random jitter so that each BIRD instance doesn't restart at once.

@matthewdupre any other specific timing things you were concerned about?

**Comment by caseydavenport** [comment]
@djosborne you did some work in this area a while back, didn't you? Do you have time this week to try this out?

**Comment by matthewdupre** [comment]
@caseydavenport couple of concerns:
- Need to make sure BIRD updates an already programmed route (since the route will survive BIRD restart).
- BGP GR can take quite a long time (default is 10s of seconds IIRC).  If we actually need to only have one restarting BIRD at once, this might be problematic and take hours in a large cluster.  I'm not sure how much overlapping GR windows is actually a problem though.
- Can BIRD 1 restart early, receive a bad (eth0) route from another not-yet-restarted BIRD, program that route and then ignore a later update (tunl0) from BIRD 2 when BIRD 2 is restarted?

**Comment by caseydavenport** [comment]
@matthewdupre Yeah, that all sounds pretty scary.

Maybe we should be looking for the BIRD fix instead.

**Comment by caseydavenport** [comment]
This has been around for a long time and it's really bad.  Moving to P0

**Comment by fasaxc** [comment]
Does this BIRD PR help at all: https://github.com/projectcalico/calico-bird/pull/25 I thought that was supposed to fix BIRD's inability to replace IPIP/non-IPIP routes with the opposite.

**Comment by caseydavenport** [comment]
@fasaxc it might!  We should investigate and see if this can just be closed when we move to that version.

**Comment by matthewdupre** [comment]
Let's hope so :).

**Comment by robbrockbank** [comment]
I've tested this using a 2 node ubuntu vagrant cluster.

```
vagrant@calico-01:~/tmp/calico-containers$ ip route
default via 10.0.2.2 dev enp0s3 
10.0.2.0/24 dev enp0s3  proto kernel  scope link  src 10.0.2.15 
172.17.8.0/24 dev enp0s8  proto kernel  scope link  src 172.17.8.101 
172.18.0.0/16 dev docker0  proto kernel  scope link  src 172.18.0.1 linkdown 
192.168.118.128 dev calia4e40b0a9a2  scope link 
blackhole 192.168.118.128/26  proto bird 
192.168.246.192/26 via 172.17.8.102 dev enp0s8  proto bird 
vagrant@calico-01:~/tmp/calico-containers$ calicoctl get pool --output=wide
CIDR                       NAT     IPIP    
192.168.0.0/16             false   false   
fd80:24e2:f998:72d6::/64   false   false   

vagrant@calico-01:~/tmp/calico-containers$ calicoctl get pool --output=yaml
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: 192.168.0.0/16
  spec: {}
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: fd80:24e2:f998:72d6::/64
  spec: {}
vagrant@calico-01:~/tmp/calico-containers$ calicoctl get pool --output=yaml > pools.yaml
vagrant@calico-01:~/tmp/calico-containers$ vim pools.yaml 
vagrant@calico-01:~/tmp/calico-containers$ calicoctl apply -f pools.yaml                
Successfully applied 2 'ipPool' resource(s)
vagrant@calico-01:~/tmp/calico-containers$ ip route
default via 10.0.2.2 dev enp0s3 
10.0.2.0/24 dev enp0s3  proto kernel  scope link  src 10.0.2.15 
172.17.8.0/24 dev enp0s8  proto kernel  scope link  src 172.17.8.101 
172.18.0.0/16 dev docker0  proto kernel  scope link  src 172.18.0.1 linkdown 
192.168.118.128 dev calia4e40b0a9a2  scope link 
blackhole 192.168.118.128/26  proto bird 
192.168.246.192/26 via 172.17.8.102 dev tunl0  proto bird onlink 
vagrant@calico-01:~/tmp/calico-containers$ ifconfig -a
```

Route moves over to tunl0 immediately.

I'll keep this issue open to see if anyone else wants to check it is fixed in other repo scenarios.

**Comment by caseydavenport** [comment]
Fixed in v1.0.0-rc1
